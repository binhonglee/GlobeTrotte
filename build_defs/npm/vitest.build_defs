subinclude("//build_defs/npm")

def vitest_dir(
  name:str,
  srcs:list=[],
  config:str=None,
  deps:list=[],
  visibility:list=[],
):
  cmd = "zip -r " + name + ".vitest_dir ./"
  if config is not None:
    cmd = " && ".join([
      "rm vite.config.ts",
      "mv " + config + " vite.config.ts",
      cmd,
    ])

  return build_rule(
    name = name,
    srcs = srcs + [config],
    cmd = cmd,
    outs = [name + ".vitest_dir"],
    deps = deps,
    test_only = True,
    visibility = visibility
  )

def vitest_tests(
  name:str,
  srcs:list=[],
  vitest_dirs:list=[],
  deps:list=[],
  visibility:list=[],
):
  test_rules = []

  for src in srcs:
    test = vitest_test(
      name = src[:-8].replace("/", "_"),
      src = src,
      vitest_dirs = vitest_dirs,
      deps = deps,
      visibility = visibility,
    )
    test_rules += [test]

  return build_rule(
    name = name,
    data = test_rules + srcs,
    test_cmd = "echo 'test'",
    test_only = True,
    visibility = visibility,
    requires = ["npm-test"],
  )

def vitest_test(
  name:str,
  src:str,
  vitest_dirs:list=[],
  deps:list=[],
  visibility:list=[],
):
  pre_build = _set_node_modules()
  if len(vitest_dirs) > 0:
    pre_build = " && ".join([
      pre_build,
      "; ".join([
        "for z in $(find . -name '*.vitest_dir' -type f)",
        "do unzip -o $z",
        "done",
      ]),
    ])

  out_dir = _vitest_out_dir(name)
  if CONFIG.CORE_FILES != "":
    deps += [CONFIG.CORE_FILES]

  return build_rule(
    name = name,
    data = [src] + vitest_dirs + deps,
    test_cmd = " && ".join([
      pre_build,
      " ".join([
        _get_npx_tool(),
        "vitest",
        "--coverage",
        "--threads=false",
        "--reporter=junit",
        "--run",
        "$PKG_DIR/" + src,
      ]),
      "cp vitest_junit test.results",
      "cp vitest_coverage/coverage-final.json test.coverage",
      "mkdir -p " + out_dir,
      "mv vitest_coverage/* " + out_dir,
    ]),
    test_outputs = [out_dir],
    test = True,
    test_only = True,
    visibility = visibility,
    labels = ['vitest'],
  )

def _vitest_out_dir(name: str):
  return name + "/.nyc_output"
