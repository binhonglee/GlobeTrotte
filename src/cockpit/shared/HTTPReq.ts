import { LoadingBarApiInjection } from "naive-ui/lib/loading-bar/src/LoadingBarProvider";
import { useLoadingBar } from "naive-ui";
import H from "./H";
import router from "@/router";
import Routes from "@/routes";

export default class HTTPReq extends H {
  protected static host =
    process.env.NODE_ENV === "production" ? "globetrotte.com" : "localhost";
  protected static port =
    process.env.NODE_ENV === "production" ? undefined : 4000;
  protected static protocol =
    process.env.NODE_ENV === "production" ? "https://" : "http://";
  protected static pathPrefix = "/api/";
  protected static delPrefix = "del/";
  protected static rateLimited = "ratelimited";
  protected static router = router;

  protected static selfPort =
    process.env.NODE_ENV === "production" ? undefined : 3000;
  private static loadingBar: LoadingBarApiInjection | undefined = undefined;
  private static failed = false;

  protected static beforeSendRequest(): void {
    try {
      this.loadingBar = useLoadingBar();
      this.loadingBar.start();
      this.failed = false;
    } catch (_) {
      this.failed = true;
    }
  }

  public static getAbsoluteURL(routes: Routes, args = ""): string {
    return (
      this.protocol +
      this.host +
      (this.selfPort === undefined ? "" : ":" + this.selfPort) +
      routes +
      "/" +
      args
    );
  }

  protected static sendRequestSuccess(): void {
    if (!this.failed && this.loadingBar !== undefined) {
      this.loadingBar.finish();
    }
  }

  protected static sendRequestFailure(): void {
    if (!this.failed && this.loadingBar !== undefined) {
      this.loadingBar.error();
    }
  }
}
