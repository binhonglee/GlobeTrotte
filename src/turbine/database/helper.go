package database

// NotExists - Return an array of int from haystack that aren't in the needles
func NotExists(haystack []int, needles []int) []int {
	needs := make(map[int]bool)
	notExists := make(map[int]bool)

	for _, v := range needles {
		needs[v] = true
	}

	for _, v := range haystack {
		if !needs[v] {
			notExists[v] = true
		}
	}
	toReturn := make([]int, len(notExists))
	i := 0
	for key, _ := range notExists {
		toReturn[i] = key
		i++
	}

	return toReturn
}
