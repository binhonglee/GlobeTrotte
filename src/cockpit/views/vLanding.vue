<template lang="pug">
  .landing
    h1.title GlobeTrotte
    p
      | Feel free to click around but nothing is set in
      | stone. Do not save any important infomation here.
      | Everything can and will be wiped from time to time.
      br
      br
      | You should check back again soon!
    br
    br
    .homePageTripCarousel
      el-card.homePageTripCard(
        v-for="trip in trips"
        shadow="hover"
      )
        CTripInCarousel(
          :trip="trip"
        )
</template>

<script lang="ts">
import CTripInCarousel from "@/components/CTripInCarousel.vue";
import HTTPReq from "@/shared/HTTPReq";
import TripObj from "@/wings/TripObj";

interface Data {
  trips: TripObj[];
}

export default {
  components: {
    CTripInCarousel,
  },
  data: (): Data => ({
    trips: [],
  }),
  async beforeMount(): Promise<void> {
    this.$data.trips = [];
    const trips = await HTTPReq.genGET(this.$router, "v2/sample_trips");
    for (const trip of trips as Array<unknown>) {
      const newTrip = new TripObj(trip);
      if (newTrip.details.days.length > 0) {
        this.$data.trips.push(new TripObj(trip));
      }
    }
  },
};
</script>

<style lang="scss">
@import "../shared/lib";

.homePageTripCarousel {
  padding: 0 0 80px 0;

  .viewUserTripInfoCarousel .el-carousel__container {
    height: 350px;
  }
}

.homePageTripCard {
  display: inline-block;
  margin: 15px;
  width: 310px;

  .el-card__body {
    padding: 5px;
  }
}
</style>
