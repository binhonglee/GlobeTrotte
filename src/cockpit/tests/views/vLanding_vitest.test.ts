import HTTPReq from "@/shared/HTTPReq";
import vLanding from "@/views/vLanding.vue";
import { expect, spyOn, test } from "vitest";
import { mount } from "@vue/test-utils";
import { mountingOptions } from "../helper";
import { stub } from "../vitestSpy";

test("Landing - Load Page", async () => {
  const genGET = stub(spyOn(HTTPReq, "genGET")).resolves([]);
  const wrapper = mount(vLanding, mountingOptions());
  expect(wrapper.html()).toMatchSnapshot();
  expect(wrapper.find(".title").text()).toEqual("GlobeTrotte");
  expect(genGET.calledOnce()).toEqual(true);
  expect(genGET.args()[0][0]).toEqual("v2/sample_trips");
  await genGET.restore();
});
