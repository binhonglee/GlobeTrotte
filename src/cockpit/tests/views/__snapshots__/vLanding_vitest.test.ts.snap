// Vitest Snapshot v1

exports[`Landing - Load Page 1`] = `
<div class="landing"
     data-v-299ea923
>
  <h1 class="title"
      data-v-299ea923
  >
    GlobeTrotte
  </h1>
  <h3 class="subtitle"
      data-v-299ea923
  >
    Look for your next travel plan here!
  </h3>
  <br data-v-299ea923>
  <form class="tripSearchForm"
        data-v-299ea923
  >
    <div class="n-input n-input--resizable n-input--stateful tripSearchQueryInput"
         style="--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-count-text-color: rgb(118, 124, 130); --n-color: rgba(255, 255, 255, 1); --n-font-size: 14px; --n-border-radius: 3px; --n-height: 34px; --n-padding-left: 12px; --n-padding-right: 12px; --n-text-color: rgb(51, 54, 57); --n-caret-color: #18a058; --n-text-decoration-color: rgb(51, 54, 57); --n-border: 1px solid rgb(224, 224, 230); --n-border-disabled: 1px solid rgb(224, 224, 230); --n-border-hover: 1px solid #36ad6a; --n-border-focus: 1px solid #36ad6a; --n-placeholder-color: rgba(194, 194, 194, 1); --n-placeholder-color-disabled: rgba(209, 209, 209, 1); --n-icon-size: 16px; --n-line-height-textarea: 1.6; --n-color-disabled: rgb(250, 250, 252); --n-color-focus: rgba(255, 255, 255, 1); --n-text-color-disabled: rgba(194, 194, 194, 1); --n-box-shadow-focus: 0 0 0 2px rgba(24, 160, 88, 0.2); --n-loading-color: #18a058; --n-caret-color-warning: #f0a020; --n-color-focus-warning: rgba(255, 255, 255, 1); --n-box-shadow-focus-warning: 0 0 0 2px rgba(240, 160, 32, 0.2); --n-border-warning: 1px solid #f0a020; --n-border-focus-warning: 1px solid #fcb040; --n-border-hover-warning: 1px solid #fcb040; --n-loading-color-warning: #f0a020; --n-caret-color-error: #d03050; --n-color-focus-error: rgba(255, 255, 255, 1); --n-box-shadow-focus-error: 0 0 0 2px rgba(208, 48, 80, 0.2); --n-border-error: 1px solid #d03050; --n-border-focus-error: 1px solid #de576d; --n-border-hover-error: 1px solid #de576d; --n-loading-color-error: #d03050; --n-clear-color: rgba(194, 194, 194, 1); --n-clear-size: 16px; --n-clear-color-hover: rgba(146, 146, 146, 1); --n-clear-color-pressed: rgba(175, 175, 175, 1); --n-icon-color: rgba(194, 194, 194, 1); --n-icon-color-hover: rgba(146, 146, 146, 1); --n-icon-color-pressed: rgba(175, 175, 175, 1); --n-icon-color-disabled: rgba(209, 209, 209, 1); --n-suffix-text-color: rgb(51, 54, 57);"
         data-v-299ea923
    >
      <div class="n-input-wrapper">
        <div class="n-input__input">
          <input type="text"
                 class="n-input__input-el"
                 style
                 placeholder="Alaska"
                 size="20"
          >
          <div class="n-input__placeholder">
            <span>
              Alaska
            </span>
          </div>
        </div>
      </div>
      <div class="n-input__border">
      </div>
      <div class="n-input__state-border">
      </div>
    </div>
    <div class="n-select tripSearchCityInput"
         data-v-299ea923
    >
      <div class="n-base-selection n-base-selection--multiple"
           style="--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-border: 1px solid rgb(224, 224, 230); --n-border-active: 1px solid #18a058; --n-border-focus: 1px solid #36ad6a; --n-border-hover: 1px solid #36ad6a; --n-border-radius: 3px; --n-box-shadow-active: 0 0 0 2px rgba(24, 160, 88, 0.2); --n-box-shadow-focus: 0 0 0 2px rgba(24, 160, 88, 0.2); --n-caret-color: #18a058; --n-color: rgba(255, 255, 255, 1); --n-color-active: rgba(255, 255, 255, 1); --n-color-disabled: rgb(250, 250, 252); --n-font-size: 14px; --n-height: 34px; --n-padding-single: 0 26px 0 12px; --n-padding-multiple: 3px 26px 0 12px; --n-placeholder-color: rgba(194, 194, 194, 1); --n-placeholder-color-disabled: rgba(209, 209, 209, 1); --n-text-color: rgb(51, 54, 57); --n-text-color-disabled: rgba(194, 194, 194, 1); --n-arrow-color: rgba(194, 194, 194, 1); --n-arrow-color-disabled: rgba(209, 209, 209, 1); --n-loading-color: #18a058; --n-color-active-warning: rgba(255, 255, 255, 1); --n-box-shadow-focus-warning: 0 0 0 2px rgba(240, 160, 32, 0.2); --n-box-shadow-active-warning: 0 0 0 2px rgba(240, 160, 32, 0.2); --n-border-warning: 1px solid #f0a020; --n-border-focus-warning: 1px solid #fcb040; --n-border-hover-warning: 1px solid #fcb040; --n-border-active-warning: 1px solid #f0a020; --n-color-active-error: rgba(255, 255, 255, 1); --n-box-shadow-focus-error: 0 0 0 2px rgba(208, 48, 80, 0.2); --n-box-shadow-active-error: 0 0 0 2px rgba(208, 48, 80, 0.2); --n-border-error: 1px solid #d03050; --n-border-focus-error: 1px solid #de576d; --n-border-hover-error: 1px solid #de576d; --n-border-active-error: 1px solid #d03050; --n-clear-size: 16px; --n-clear-color: rgba(194, 194, 194, 1); --n-clear-color-hover: rgba(146, 146, 146, 1); --n-clear-color-pressed: rgba(175, 175, 175, 1); --n-arrow-size: 16px;"
      >
        <div class="n-base-selection-tags"
             tabindex="0"
        >
          <div class="n-base-selection-input-tag">
            <input tabindex="-1"
                   class="n-base-selection-input-tag__input"
            >
            <span class="n-base-selection-input-tag__mirror">
            </span>
          </div>
          <div class="n-base-loading n-base-suffix"
               role="img"
               aria-label="loading"
          >
            <transition-stub>
              <div class="n-base-loading__placeholder">
                <div class="n-base-clear">
                  <transition-stub>
                    <div class="n-base-clear__placeholder">
                      <i class="n-base-icon n-base-suffix__arrow">
                        <svg viewbox="0 0 16 16"
                             fill="none"
                             xmlns="http://www.w3.org/2000/svg"
                        >
                          <path d="M3.14645 5.64645C3.34171 5.45118 3.65829 5.45118 3.85355 5.64645L8 9.79289L12.1464 5.64645C12.3417 5.45118 12.6583 5.45118 12.8536 5.64645C13.0488 5.84171 13.0488 6.15829 12.8536 6.35355L8.35355 10.8536C8.15829 11.0488 7.84171 11.0488 7.64645 10.8536L3.14645 6.35355C2.95118 6.15829 2.95118 5.84171 3.14645 5.64645Z"
                                fill="currentColor"
                          >
                          </path>
                        </svg>
                      </i>
                    </div>
                  </transition-stub>
                </div>
              </div>
            </transition-stub>
          </div>
        </div>
        <div class="n-base-selection-placeholder n-base-selection-overlay">
          City
        </div>
        <div class="n-base-selection__border">
        </div>
        <div class="n-base-selection__state-border">
        </div>
      </div>
    </div>
    <button class="n-button n-button--default-type n-button--medium-type"
            tabindex="0"
            type="button"
            style="--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-bezier-ease-out: cubic-bezier(0, 0, .2, 1); --n-ripple-duration: .6s; --n-opacity-disabled: 0.5; --n-wave-opacity: 0.6; font-weight: 400; --n-color: #0000; --n-color-hover: #0000; --n-color-pressed: #0000; --n-color-focus: #0000; --n-color-disabled: #0000; --n-ripple-color: #18a058; --n-text-color: rgb(51, 54, 57); --n-text-color-hover: #36ad6a; --n-text-color-pressed: #0c7a43; --n-text-color-focus: #36ad6a; --n-text-color-disabled: rgb(51, 54, 57); --n-border: 1px solid rgb(224, 224, 230); --n-border-hover: 1px solid #36ad6a; --n-border-pressed: 1px solid #0c7a43; --n-border-focus: 1px solid #36ad6a; --n-border-disabled: 1px solid rgb(224, 224, 230); --n-width: initial; --n-height: 34px; --n-font-size: 14px; --n-padding: 0 14px; --n-icon-size: 18px; --n-icon-margin: 6px; --n-border-radius: 3px;"
            data-v-299ea923
    >
      <transition-stub>
      </transition-stub>
      <span class="n-button__content">
        Find
      </span>
      <div aria-hidden="true"
           class="n-base-wave"
      >
      </div>
      <div aria-hidden="true"
           class="n-button__border"
      >
      </div>
      <div aria-hidden="true"
           class="n-button__state-border"
      >
      </div>
    </button>
  </form>
  <div class="homePageTripCarousel"
       data-v-299ea923
  >
  </div>
</div>
`;
